plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -332,6 +355,16 @@ private void setFieldValue(String path, Object value, boolean append) {
         throw new IllegalArgumentException("field [" + path + "] of type [" + object.getClass().getName() + "] cannot be cast to [" + clazz.getName() + "]");
     }
 
+    private Map<String, Object> createTemplateModel() {
+        Map<String, Object> model = new HashMap<>(sourceAndMetadata);
+        // TODO: I think we should split source from meta data? because now also ES metadata is under the _source key...
+        model.put("_source", sourceAndMetadata);
+        // If there is a field in the source with the name '_ingest' it gets overwritten here,
+        // if access to that field is required then it get accessed via '_source._ingest'
+        model.put("_ingest", ingestMetadata);
+        return model;
```

Comment:
looking for a way to have the same logic that we already have in IngestDocument here too... the behaviour needs to be unified either via templates or via getFieldValue. I am not sure that splitting source and metadata again is the solution, cause it will complicate things again on the getFieldValue/setFieldValue side of things. need to think about this some more.